src/maps/hashmap.js

```diff
@@ -0,0 +1,134 @@
+const IAMap = require('iamap')
+const murmurhash3 = require('murmurhash3js-revisited')
+const Type = require('../base')
+const Block = require('@ipld/block')
+
+const _type = 'IPLD/Experimental/HashMap/0'
+
+class HashMap extends Type {
+  get _type () {
+    return _type
+  }
+
+  get kind () {
+    return 'map'
+  }
+
+  async get (args, continuation) {
+    let path = args.path.split('/').filter(Boolean)
+    let key = path.shift()
+
+    if (!key) {
+      throw new Error('HashMap type needs a string to look up')
+    }
+
+    // TODO: this should be a validation of this.node against the HAMT root block schema
+    if (!IAMap.isRootSerializable(this.node) || !IAMap.isSerializable(this.node)) {
+      throw new Error('Block does not contain expected HashMap root data')
+    }
+
+    // provide a store to IAMap that won't do any retrieval but will intercept load() calls
+    // so we can pass that up for continuation
+    let target
+    const fauxStorage = {
+      save (obj) {
+        throw new Error('Mutations are not supported during get() operations')
+      },
+      load (id) {
+        if (target) {
+          throw new Error('Single load() expected during traversal')
+        }
+        target = id
+        return null
+      },
+      isEqual (cid1, cid2) {
+        return cid1.equals(cid2)
+      }
+    }
+
+    // instantiate a root IAMap node
+    let root = IAMap.fromSerializable(fauxStorage, null, this.node)
+    let map = root
+    let depth = 0
+    if (continuation) {
+      // we have navigated to a child node of with depth of continuation.depth
+      if (!IAMap.isSerializable(continuation.response.node)) {
+        throw new Error('Block does not contain expected HashMap child node data')
+      }
+      depth = continuation.depth + 1
+      // instantiate a child node based on the config of the root node but at `depth` depth
+      map = root.fromChildSerializable(null, continuation.response.node, depth)
+    }
+
+    try {
+      let result = await map.get(key)
+      if (!path.length) {
+        return { result }
+      }
+      throw new Error('Traversing into values is not yet supported')
+      // what do? could be an arbitrary kind
+      // return { call: { info: { method: 'get', args: path.join('/') }, target: result, proxy: true } }
```

Comment:
I don't think this is supported yet but I'd like to be able to return an arbitrary thing and be able to navigate into it, e.g. the map could store objects inline and you should be able to skip into those objects via remaining parts of `path`, see tests for examples of what I think you should be able to do.

Reply:
you should already be able to do the thing you’ve commented out. this works for `get` because the path is an arg. any target you return that isn’t a string is treated as the target directly, this should work. it end up not working in the non-`get` case if you need (target + path) because other methods won’t resolve the path as an arg and you only get one thing you’re allowed to return as the target.